### ISSUE TITLE
Make new type for programs

### ISSUE THREAD

**Issue by Danten** [advancedtelematic/quickcheck-state-machine-model#30]
Instead of having a program be of type `[Untyped cmd ref]` we could have a new type
```
data Program cmd ref where
  Done :: Program cmd ref
  Let :: IntRef -> cmd ('Reference ix) ref -> Program cmd ref -> Program cmd ref
  Action :: cmd ('Response t) ref -> Program cmd ref -> Program cmd ref
```
I'm not yet sure if having two different constructors `Let` and `Action` is worth it yet, in case one more often makes a case distinction on the result of `returns` then it wouldn't be that bad.

Either way the important thing is the `IntRef` in the `Let` constructor, so instead of:
- Generating the new variables in `liftSem`
- Guess (correctly) the names that will be generated in `liftGen`
- Remap all variable names in the `liftShrink`

We make `liftGen` add the names (and they are now really names and not just be a de Bruijn level), and then have:
- `liftSem` just uses that name when it adds things to the environment
- `liftShrink` only cares about removing commands that uses the reference (no need to re-index names)

**Comment by Danten** [comment]
Oh yeah, the `fixRefs` replacement would be something like this:
```
uses :: IxFoldable (cmd resp) => cmd resp (Pair IntRef) -> Set IntRef -> cmd resp (Pair IntRef)
uses cmd xs = anyIx (\ _ (_, ref) -> ref `S.elem` xs) cmd

(\\) :: Program cmd (Pair IntRef) -> Set IntRef -> Program cmd (Pair IntRef)
Done \\ _ = Done
Let x cmd prog \\ xs | uses cmd xs = prog \\ (S.insert x xs)
                     | otherwise = Let x cmd (prog \\ xs)
Action cmd prog \\ xs | uses cmd xs = prog \\ xs
                      | otherwise = Action cmd (prog \\ xs)
```

**Comment by stevana** [comment]
Implemented in a slightly different way as part of #33.
